### ISSUE TITLE
Changing query parameters for GA4

### ISSUE THREAD

**Issue by mikestreety** [tpwd/ke_search#124]
Hey,

This isn't an issue with KE Search _per se_, but it affects it and was wondering if we could hash out a solution.

We are changing to GA4 and, in doing so, Google have prevented square brackets `[ ]` from being entered into the site search tracking section, which means we can no longer get detailed reports on site search 😞 

I got in touch with Google support and they confirmed this.

Is there a way I, as an extension user, can change the parameter in the URL? I know it will kind of go against TYPO3 conventions, but didn't know if I could use a URL enhancer in the site config or something to make it `query={sword}`. I tried, but couldn't get it to work.

Thanks for any assistance on this 👍

**Comment by christianbltr** [comment]
Hey,

yes, you can use routing with speaking URLs as described here:

https://docs.typo3.org/p/tpwd/ke_search/main/en-us/Configuration/RoutingSpeakingUrls.html

Does that help you out?

**Comment by mikestreety** [comment]
Hey,

I tried to use it but for some reason I couldn't get it to stick. 🤔 It just wouldn't work for me. 

Maybe I'll try again and drop my code in this issue 👍

**Comment by christianbltr** [comment]
That would be great.

I just tested this simple example (without any filters) in TYPO3 11.5.16 (just copy it to your site config) and it worked (make sure to clear the lowlevel cache at "Maintenance" --> "Flush TYPO3 and PHP Cache").

```
routeEnhancers:
  KeSearch:
    type: Plugin
    routePath: '{sortByField}/{sortByDir}/{resetFilters}/{page}/{sword}'
    namespace: 'tx_kesearch_pi1'
    defaults:
      sortByField: 'score'
      sortByDir: 'desc'
      resetFilters: '0'
      page: '1'
      sword: ''
      filter_13: ''
    requirements:
      sortByField: '(score|title|customranking|sortdate)?'
      sortByDir: '(asc|desc)?'
      resetFilters: '[0-9]?'
      page: '\d+'
    aspects:
      sortByField:
        type: StaticValueMapper
        map:
          score: 'score'
          customranking: 'customranking'
          title: 'title'
      sortByDir:
        type: StaticValueMapper
        map:
          asc: 'asc'
          desc: 'desc'
      resetFilters:
        type: StaticRangeMapper
        start: '0'
        end: '1'
      page:
        type: StaticRangeMapper
        start: '1'
        end: '99'
      sword:
        type: KeSearchUrlEncodeMapper
```

**Comment by mikestreety** [comment]
Hmm, still couldn't get this to work with `?query=` - I ended up doing it with a couple of `XClasses`.

Both analytics and schema flag up the square brackets as invalid

**Comment by christianbltr** [comment]
Ah, sorry, I think I overlooked that you wanted to use a GET parameter `query` like in `?query={sword}`. The routing configuration includes the searchword as part of the path, not as a parameter.

Would you mind sharing your solution? Maybe that would be something that can be included in ke_search itself as a configurable alternative to the default parameter for the search word.

**Comment by mikestreety** [comment]
@christianbltr Of course, see below (hopefully it makes sense).

**`setup.typoscript`**

Allows the item to be customised in TypoScript & retrieved by code

```
plugin.tx_kesearch_pi1 {
	parameter = query
}
```

**`Resources/Private/Templates/SearchForm.html`**

```diff
<div class="kesearchbox">
-	<input type="search" id="ke_search_sword" name="tx_kesearch_pi1[sword]" value="{searchword -> f:format.raw()}" placeholder="{searchwordDefault}" {f:if(condition: '{extConf.allowEmptySearch} == 0', then: 'required')} minlength="{extConf.searchWordLength}" />
+	<input type="search" id="ke_search_sword" name="{conf.parameter}" value="{searchword -> f:format.raw()}" placeholder="{searchwordDefault}" {f:if(condition: '{extConf.allowEmptySearch} == 0', then: 'required')} minlength="{extConf.searchWordLength}" />
	<div class="clearer">&nbsp;</div>
</div>
```

**`Classes/Lib/Pluginbase.php`**

```php
/**
 * Initializes flexform, conf vars and some more
 * @return void
 */
public function init()
{
	// Use "query=" in URL but map to tx_kesearch_pi1[sword]=
	$param = $GLOBALS['TSFE']->tmpl->setup['site.']['search.']['parameter'];
	if (!isset($this->piVars['sword']) && GeneralUtility::_GP($param)) {
		$this->piVars['sword'] = GeneralUtility::_GP($param);
	}

	// existing code
}
```

**Comment by christianbltr** [comment]
Thanks.

This feature is now merged to master and will be in the next release.

The search word parameter can be changed with this Setup TypoScript setting:

```
plugin.tx_kesearch_pi1.searchWordParameter = query
```

**Comment by mikestreety** [comment]
Amazing, thanks @christianbltr!

**Comment by quadronet** [comment]
In which version is this function integrated? 

plugin.tx_kesearch_pi1.searchWordParameter = query

Thank´s

**Comment by christianbltr** [comment]
That is available since version 4.6.0, see https://github.com/tpwd/ke_search/blob/master/ChangeLog#L102
